examples/text/plot_document_clustering.py

```diff
@@ -162,126 +46,187 @@ def is_interactive():
 
-print("Extracting features from the training dataset using a sparse vectorizer")
-t0 = time()
-if opts.use_hashing:
-    if opts.use_idf:
-        # Perform an IDF normalization on the output of HashingVectorizer
-        hasher = HashingVectorizer(
-            n_features=opts.n_features,
-            stop_words="english",
-            alternate_sign=False,
-            norm=None,
-        )
-        vectorizer = make_pipeline(hasher, TfidfTransformer())
-    else:
-        vectorizer = HashingVectorizer(
-            n_features=opts.n_features,
-            stop_words="english",
-            alternate_sign=False,
-            norm="l2",
-        )
-else:
     vectorizer = TfidfVectorizer(
         max_df=0.5,
-        max_features=opts.n_features,
+        max_features=10000,
         min_df=2,
         stop_words="english",
-        use_idf=opts.use_idf,
+        use_idf=idf,
     )
-X = vectorizer.fit_transform(dataset.data)
 
-print("done in %fs" % (time() - t0))
-print("n_samples: %d, n_features: %d" % X.shape)
-print()
+    X = vectorizer.fit_transform(data)
 
-if opts.n_components:
-    print("Performing dimensionality reduction using LSA")
-    t0 = time()
-    # Vectorizer results are normalized, which makes KMeans behave as
-    # spherical k-means for better results. Since LSA/SVD results are
-    # not normalized, we have to redo the normalization.
-    svd = TruncatedSVD(opts.n_components)
-    normalizer = Normalizer(copy=False)
-    lsa = make_pipeline(svd, normalizer)
+    pipeline.append(vectorizer)
 
-    X = lsa.fit_transform(X)
+    if type(n_components) is int:
+        # Vectorizer results are normalized, which makes KMeans behave as
+        # spherical k-means for better results. Since LSA/SVD results are
+        # not normalized, we have to redo the normalization.
+        svd = TruncatedSVD(10, random_state=random_state)
+        normalizer = Normalizer(copy=False)
+        lsa = make_pipeline(svd, normalizer)
 
-    print("done in %fs" % (time() - t0))
+        X = lsa.fit_transform(X)
 
-    explained_variance = svd.explained_variance_ratio_.sum()
-    print(
-        "Explained variance of the SVD step: {}%".format(int(explained_variance * 100))
-    )
+        pipeline.append(svd)
 
-    print()
-
-
-# %%
-# Clustering
-# ----------
-
-if opts.minibatch:
     km = MiniBatchKMeans(
-        n_clusters=true_k,
+        n_clusters=n_clusters,
         init="k-means++",
         n_init=1,
         init_size=1000,
         batch_size=1000,
-        verbose=opts.verbose,
-    )
-else:
-    km = KMeans(
-        n_clusters=true_k,
-        init="k-means++",
-        max_iter=100,
-        n_init=1,
-        verbose=opts.verbose,
+        verbose=False,
+        random_state=random_state,
     )
 
-print("Clustering sparse data with %s" % km)
-t0 = time()
-km.fit(X)
-print("done in %0.3fs" % (time() - t0))
-print()
+    km.fit(X)
+
+    pipeline.append(km)
+
+    homo = metrics.homogeneity_score(labels, km.labels_)
```

Comment:
I think that we can isolate the evaluation in its own function:

```python
from collections import defaultdict

import numpy as np
import pandas as pd

from sklearn.metrics import (
    homogeneity_score,
    completeness_score,
    v_measure_score,
    adjusted_rand_score,
    silhouette_score,
)


def fit_evaluate_clusterer_booststrap(clusterer, X, labels, n_bootstrap=10, random_state=None):
    bootstrap_results = defaultdict(list)
    n_samples = X.shape[0]
    rng = np.random.RandomState(random_state)

    for _ in range(n_bootstrap):
        bootstrap_indices = rng.choice(np.arange(n_samples), size=n_samples, replace=True)
        X_bootstrap, labels_bootstrap = X[bootstrap_indices], labels[bootstrap_indices]
        predicted_labels = clusterer.fit_predict(X_bootstrap)
    
        bootstrap_results["homogeneity"].append(homogeneity_score(labels_bootstrap, predicted_labels))
        bootstrap_results["completness"].append(completeness_score(labels_bootstrap, predicted_labels))
        bootstrap_results["v_measure"].append(v_measure_score(labels_bootstrap, predicted_labels))
        bootstrap_results["ARI"].append(adjusted_rand_score(labels_bootstrap, predicted_labels))
        bootstrap_results["silhouhette"].append(silhouette_score(X_bootstrap, predicted_labels, sample_size=20))
    
    return pd.DataFrame(bootstrap_results, index=[f"Bootstrap #{i}" for i in range(n_bootstrap)])
```

The idea of using the following function will be to get an estimate of the variation of the fitting using bootstrap. We had a discussion with @ogrisel IRL to know what would be the best strategy here and it seems the best.

We can then call this function and show the mean and std results:

```python
bootstrap_results = fit_evaluate_clusterer_booststrap(kmeans, X, y, 10, np.random.RandomState(0))
bootstrap_results.aggregate(["mean", "std"])
```

Since we will have 3 different models, we can create the dataframe a bit differently but showing the method as index and the metric/mean/std with a multi-index as columns could be a way to summarize the information nicely.

Reply:
Thanks for the detailed review.

But I encountered an issue. Silhouette coeff takes as argument the data in feature space. `X_bootstrap` in the function is a list of the documents. Which means we need the output of the vectorizer but `clusterer` is a `Pipeline` and doesn't output intermediary layers (unless there is a way I'm not aware of). Same problem with the last code cell where the plotting happens and all the steps are needed.

So we either remove silhouette coef, or create pipelines only for the feature extraction and instantatiate `KMeans` separately. Any suggestions?